site/content/blog/multi-env-deployment-ytt-kapp.md

```diff
@@ -0,0 +1,331 @@
+  replicas: #@ data.values.replicaCount
+  selector:
+    matchLabels: #@ labels()
+  template:
+    metadata:
+      labels: #@ labels()
+    spec:
+      containers:
+        - name: sample-app
+          image: #@ data.values.image.name + ":" + data.values.image.tag
+          imagePullPolicy: IfNotPresent
+          ports:
+          - containerPort: 8080
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: application-settings
+data:
+  app_mode: #@ data.values.appMode
+  certificates: #@ data.values.certificatePath
+  db_user: #@ data.values.databaseUser
+  db_password: #@ data.values.databasePassword
+```
+
+As you can see above file contains lots of ytt annotations (i.e. lines that contain #@), it’s a ytt template. With the help of this annotations we are using the values defined in schema file. 
+
+## Deploying with kapp 
+Let's now look into how we can deploy for different environments by passing environment specific `values.yaml` we created above. First, let's see the final manifest for staging environment.
+
+```bash
+$ ytt -f app.yaml -f schema.yaml -f values-staging.yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: sample-app-staging
+  labels:
+    app.kubernetes.io/name: sample-app-staging
+spec:
+  selector:
+    app: sample-app-staging
+  ports:
+  - protocol: TCP
+    port: 80
+    targetPort: 8080
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: sample-app-staging
+  labels:
+    app.kubernetes.io/name: sample-app-staging
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: sample-app-staging
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: sample-app-staging
+    spec:
+      containers:
+      - name: sample-app
+        image: nginx:latest
+        imagePullPolicy: IfNotPresent
+        ports:
+        - containerPort: 8080
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: application-settings
+data:
+  app_mode: staging
+  certificates: /etc/ssl/staging
+  db_user: staging-user
+  db_password: staging-password
+```
+
+As our final manifest looks good let's deploy it with kapp - 
+
+```bash
+$ kapp deploy -a sample-app -f <(ytt -f app.yaml -f schema.yaml -f values-staging.yaml)
+Target cluster 'https://127.0.0.1:57418' (nodes: minikube)
+
+Changes
+
+Namespace  Name                  Kind        Conds.  Age  Op      Op st.  Wait to    Rs  Ri  
+default    application-settings  ConfigMap   -       -    create  -       reconcile  -   -  
+^          sample-app-staging    Deployment  -       -    create  -       reconcile  -   -  
+^          sample-app-staging    Service     -       -    create  -       reconcile  -   -  
+
+Op:      3 create, 0 delete, 0 update, 0 noop, 0 exists
+Wait to: 3 reconcile, 0 delete, 0 noop
+
+Continue? [yN]: 
+```
+
+Here [kapp](https://carvel.dev/kapp/) is showing calculated changes between configuration provided and live cluster state and asking for confirmation before actually applying the change.
```

Comment:
```suggestion
Here [kapp](https://carvel.dev/kapp/) is showing calculated changes between configuration provided and live cluster state. It then asks for confirmation before actually applying the change.
```

nit: the first "and" connects the sources for the calculation. The subsequent phrase describes the next step. If we make this a separate sentence, that intention is clearer.

Reply:
Sounds good. Adopted.